### PR TITLE
Add transaction constraints module

### PR DESCRIPTION
**Pull request by jhbertra**
Adds definitions, documentation, and API stubs for a Marlowe-specific transaction constraints module.

Future work:
  - Implement `satisfiesConstraints` @paluh
  - Implement `solveConstraints` @dino-iog
  - Implement `buildCreateConstraints` @paluh @bjornkihlberg
  - Implement `buildApplyInputsConstraints` @paluh @bjornkihlberg
  - Implement `buildWithdrawConstraints` @paluh @bjornkihlberg
  - Implement necessary support queries (fetch UTXOs for addresses, fetch application and payout script UTXOs) @jhbertra

### CODE REVIEW

**Comment by jhbertra** [comment]
@paluh, since @dino-iog is going to be working on the constraint solver, and the constraint checker is essentially intended as a test fixture for the solver, it would probably make sense for the two of you to work on it together.

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Constraints.hs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+data InputConstraint v
+  = NoInput
+  -- ^ Specifies that the transaction consumes no marlowe or payout script
+  -- inputs. Corresponds to the create transaction type.
+  --
+  -- Rules to check:
+  --   1. All inputs of the transaction do not come from any of the script
+  --      addresses (marlowe or payout) associated with marlowe version v.
+  | MarloweInput Chain.TxOutRef P.POSIXTime P.POSIXTime (Core.Redeemer v)
```
Comment (bwbush) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Constraints.hs:141]: Should this include the datum? or will that be looked up from history?

**Review by bwbush — APPROVED** [review]
Looks good. I imagine we might need some small refinements during implementation, but I'm not sure of specifics.

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Constraints.hs (hunk 1 above)
Comment (jhbertra) [reply marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Constraints.hs:141]: Do we need the datum to construct the input? In principle, this should be computable from the existing output, so specifying it here may be redundant.

**Pull request merged**